homeassistant/components/matrix/config_flow.py

```diff
@@ -0,0 +1,131 @@
+"""Config flow for the Matrix integration."""
+import json
+import os
+from typing import Any, Dict, Final, Optional
+
+from matrix_client.client import MatrixClient
+from matrix_client.errors import MatrixHttpLibError, MatrixRequestError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import (
+    CONF_ACCESS_TOKEN,
+    CONF_BASE,
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+
+from . import MatrixAuthentication
+from .const import CONF_COMMANDS, CONF_HOMESERVER, CONF_ROOMS, DOMAIN, SESSION_FILE
+
+
+async def validate_input(hass: HomeAssistant, data: Dict[str, Any]) -> Dict[str, str]:
+    """
+    Validate the user input allows us to log in.
+
+    :return: A dict containing the title and the access token corresponding to
+    the settings provided by the input data.
+    :raises MatrixHttpLibError: if the connection fails
+    :raises MatrixRequestError: if login fails
+    """
+
+    try:
+        auth: MatrixAuthentication = MatrixAuthentication(
+            config_file=os.path.join(hass.config.path(), SESSION_FILE),
+            homeserver=data[CONF_HOMESERVER],
+            verify_ssl=data[CONF_VERIFY_SSL],
+            username=data[CONF_USERNAME],
+            password=data[CONF_PASSWORD],
+        )
+
+        client: MatrixClient = await hass.async_add_executor_job(auth._login)
+
+        # If no exception is thrown during logging in
+        token: str = ""
+        if hasattr(client, "token"):
+            # A new token will be assigned on first login
+            token = client.token
+        else:
+            # Use the token from the previous login if there is no token from the client
+            token = auth._auth_tokens[data[CONF_USERNAME]]
+
+        return {
+            "title": data[CONF_USERNAME],
+            CONF_ACCESS_TOKEN: token,
+        }
+
+    except (MatrixHttpLibError, MatrixRequestError):
+        # Re-raise so async_step_user can handle the exception
+        raise
+
+
+class ConfigFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
```

Comment:
An import is missing: `async_step_import` to import existing YAML configuration.